src/liboslexec/batched_llvm_gen.cpp

```diff
@@ -7230,14 +7231,16 @@ LLVMGEN(llvm_gen_closure)
         // For the weighted closures, we need a surrounding "if" so that it's safe
         // for osl_allocate_weighted_closure_component to return NULL (unless we
         // know for sure that it's constant weighted and that the weight is
-        // not zero).
+        // not zero).  We also need the surrounding "if" to reject writing
+        // to closures that aren't masked.
         llvm::BasicBlock* next_block    = NULL;
         llvm::BasicBlock* notnull_block = rop.ll.new_basic_block(
-            "non_null_closure");
-        next_block = rop.ll.new_basic_block("");
-        llvm::Value* cond
-            = rop.ll.op_ne(rop.ll.ptr_cast(comp_ptr, rop.ll.type_void_ptr()),
-                           rop.ll.void_ptr_null());
+            "non_null_and_masked_closure");
+        next_block        = rop.ll.new_basic_block("");
+        llvm::Value* cond = rop.ll.op_and(
+            rop.ll.op_ne(rop.ll.ptr_cast(comp_ptr, rop.ll.type_void_ptr()),
+                         rop.ll.void_ptr_null()),
+            rop.ll.test_mask_lane(mask, rop.ll.constant(lane_index)));
```

Comment:
" rop.ll.constant(lane_index)" isn't wrong, but you should be able to pass lane_index in directly without creating the constant.